File: clients/go/codegame/server_message.go
package codegame

import "fmt"
import "io"
import . "spb_ai_champ/model"
import . "spb_ai_champ/stream"

// Message sent from server
type ServerMessage interface {
    // Write ServerMessage to writer
    Write(writer io.Writer)

    // Get string representation of ServerMessage
    String() string
}

// Read ServerMessage from reader
func ReadServerMessage(reader io.Reader) ServerMessage {
    switch ReadInt32(reader) {
    case 0:
        return ReadServerMessageGetAction(reader)
    case 1:
        return ReadServerMessageFinish(reader)
    case 2:
        return ReadServerMessageDebugUpdate(reader)
    }
    panic("Unexpected tag value")
}

// Get action for next tick
type ServerMessageGetAction struct {
    // Player's view
    PlayerView Game
    // Whether app is running with debug interface available
    DebugAvailable bool
}

func NewServerMessageGetAction(playerView Game, debugAvailable bool) ServerMessageGetAction {
    return ServerMessageGetAction {
        PlayerView: playerView,
        DebugAvailable: debugAvailable,
    }
}

// Read GetAction from reader
func ReadServerMessageGetAction(reader io.Reader) ServerMessageGetAction {
    var playerView Game
    playerView = ReadGame(reader)
    var debugAvailable bool
    debugAvailable = ReadBool(reader)
    return ServerMessageGetAction {
        PlayerView: playerView,
        DebugAvailable: debugAvailable,
    }
}

// Write GetAction to writer
func (serverMessageGetAction ServerMessageGetAction) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    playerView := serverMessageGetAction.PlayerView
    playerView.Write(writer)
    debugAvailable := serverMessageGetAction.DebugAvailable
    WriteBool(writer, debugAvailable)
}

// Get string representation of GetAction
func (serverMessageGetAction ServerMessageGetAction) String() string {
    stringResult := "{ "
    stringResult += "PlayerView: "
    playerView := serverMessageGetAction.PlayerView
    stringResult += playerView.String()
    stringResult += ", "
    stringResult += "DebugAvailable: "
    debugAvailable := serverMessageGetAction.DebugAvailable
    stringResult += fmt.Sprint(debugAvailable)
    stringResult += " }"
    return stringResult
}

// Signifies end of the game
type ServerMessageFinish struct {
}

func NewServerMessageFinish() ServerMessageFinish {
    return ServerMessageFinish {
    }
}

// Read Finish from reader
func ReadServerMessageFinish(reader io.Reader) ServerMessageFinish {
    return ServerMessageFinish {
    }
}

// Write Finish to writer
func (serverMessageFinish ServerMessageFinish) Write(writer io.Writer) {
    WriteInt32(writer, 1)
}

// Get string representation of Finish
func (serverMessageFinish ServerMessageFinish) String() string {
    stringResult := "{ "
    stringResult += " }"
    return stringResult
}

// Debug update
type ServerMessageDebugUpdate struct {
    // Player's view
    PlayerView Game
}

func NewServerMessageDebugUpdate(playerView Game) ServerMessageDebugUpdate {
    return ServerMessageDebugUpdate {
        PlayerView: playerView,
    }
}

// Read DebugUpdate from reader
func ReadServerMessageDebugUpdate(reader io.Reader) ServerMessageDebugUpdate {
    var playerView Game
    playerView = ReadGame(reader)
    return ServerMessageDebugUpdate {
        PlayerView: playerView,
    }
}

// Write DebugUpdate to writer
func (serverMessageDebugUpdate ServerMessageDebugUpdate) Write(writer io.Writer) {
    WriteInt32(writer, 2)
    playerView := serverMessageDebugUpdate.PlayerView
    playerView.Write(writer)
}

// Get string representation of DebugUpdate
func (serverMessageDebugUpdate ServerMessageDebugUpdate) String() string {
    stringResult := "{ "
    stringResult += "PlayerView: "
    playerView := serverMessageDebugUpdate.PlayerView
    stringResult += playerView.String()
    stringResult += " }"
    return stringResult
}

File: clients/python/codegame/server_message.py
from model.game import Game
from stream_wrapper import StreamWrapper

class ServerMessage:
    """Message sent from server"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "ServerMessage":
        """Read ServerMessage from input stream
        """
        tag = stream.read_int()
        if tag == GetAction.TAG:
            return ServerMessage.GetAction.read_from(stream)
        if tag == Finish.TAG:
            return ServerMessage.Finish.read_from(stream)
        if tag == DebugUpdate.TAG:
            return ServerMessage.DebugUpdate.read_from(stream)
        raise Exception("Unexpected tag value")

class GetAction(ServerMessage):
    """Get action for next tick"""

    TAG = 0

    __slots__ = ("player_view","debug_available",)

    player_view: Game
    debug_available: bool

    def __init__(self, player_view: Game, debug_available: bool):
        self.player_view = player_view
        """Player's view"""
        self.debug_available = debug_available
        """Whether app is running with debug interface available"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "GetAction":
        """Read GetAction from input stream
        """
        player_view = Game.read_from(stream)
        debug_available = stream.read_bool()
        return GetAction(player_view, debug_available)
    
    def write_to(self, stream: StreamWrapper):
        """Write GetAction to output stream
        """
        stream.write_int(self.TAG)
        self.player_view.write_to(stream)
        stream.write_bool(self.debug_available)
    
    def __repr__(self):
        return "GetAction(" + \
            repr(self.player_view) + \
            ", " + \
            repr(self.debug_available) + \
            ")"

ServerMessage.GetAction = GetAction

class Finish(ServerMessage):
    """Signifies end of the game"""

    TAG = 1

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Finish":
        """Read Finish from input stream
        """
        return Finish()
    
    def write_to(self, stream: StreamWrapper):
        """Write Finish to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "Finish(" + \
            ")"

ServerMessage.Finish = Finish

class DebugUpdate(ServerMessage):
    """Debug update"""

    TAG = 2

    __slots__ = ("player_view",)

    player_view: Game

    def __init__(self, player_view: Game):
        self.player_view = player_view
        """Player's view"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugUpdate":
        """Read DebugUpdate from input stream
        """
        player_view = Game.read_from(stream)
        return DebugUpdate(player_view)
    
    def write_to(self, stream: StreamWrapper):
        """Write DebugUpdate to output stream
        """
        stream.write_int(self.TAG)
        self.player_view.write_to(stream)
    
    def __repr__(self):
        return "DebugUpdate(" + \
            repr(self.player_view) + \
            ")"

ServerMessage.DebugUpdate = DebugUpdate